### PR TITLE
rsys-1487 -- added gyro interface to robot controllers

### PR DESCRIPTION
**Pull request by navyenzo**
Ticket:
https://fetchrobotics.atlassian.net/browse/RSYS-1487

## Description
Need to provide controller direct access to gyro data though controller interface.

## Reason
The gyro will be used to stabilize the wheel controller in situations when wheels start slipping or when wheels have to go over bumps

## Tests
Code compiles but will need to be tested

### CODE REVIEW

File: robot_controllers_interface/include/robot_controllers_interface/gyro_handle.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  virtual std::string getName() = 0;
+
+  /** @brief Get the name of this joint. */
+  virtual bool isGyroDataCurrentlyValid()const = 0;
+
+
+
+protected:
+  
+  bool isGyroDataCurrentlyValid_;
```
Comment (dbking77) [inline robot_controllers_interface/include/robot_controllers_interface/gyro_handle.h:94]: ROS C++ standard uses underscores for variables and camelCase for methods and classes.
http://wiki.ros.org/CppStyleGuide#Member_variables

is_gyro_data_currently_valid_;

File: robot_controllers_interface/include/robot_controllers_interface/gyro_handle.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  std::string type_;
+  std::string name_;
+  geometry_msgs::Vector3 angularVelocities_;
```
Comment (dbking77) [inline robot_controllers_interface/include/robot_controllers_interface/gyro_handle.h:97]: angular_velocities_

File: robot_controllers_interface/include/robot_controllers_interface/gyro_handle.h [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+  /** @brief Get the name of this gyro */
+  virtual std::string getName() = 0;
+
+  /** @brief Get the name of this joint. */
+  virtual bool isGyroDataCurrentlyValid()const = 0;
+
+
+
+protected:
```
Comment (dbking77) [inline robot_controllers_interface/include/robot_controllers_interface/gyro_handle.h:92]: This should be a pure interface that only specifies what functions a gyro interface must provide.   The implementation should not have these member variables.    
For example, if getAngularVelocities() would read data directly from hardware for a register, these member variables wouldn't be needed.


See :
https://github.com/fetchrobotics/robot_controllers/blob/melodic-devel/robot_controllers_interface/include/robot_controllers_interface/joint_handle.h#L107

File: robot_controllers_interface/include/robot_controllers_interface/gyro_handle.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  /** @brief Ensure proper cleanup with virtual destructor. */
+  virtual ~GyroHandle()
+  {
+  }
+
+  /** @brief Get the angular velocities from the gyro */
+  virtual const geometry_msgs::Vector3& getAngularVelocities()const = 0;
+
+  /** @brief Set the angular velocities of the gyro */
+  virtual void setAngularVelocities(const geometry_msgs::Vector3& angularVelocities, bool isGyroDataCurrentlyValid) = 0;
```
Comment (dbking77) [inline robot_controllers_interface/include/robot_controllers_interface/gyro_handle.h:70]: For the joint handle interface, the is  get and set functions.  But they are from the controller perspective.  The controller will **get** the measured velocities and **sets** the target velocities. 
In this case, the controller can only **get** gyro angular velocties, it doesn't make sense for it to set them.
The implementation of this interface should have these **set** functions and member variables.   This interface should only provide getters.

File: robot_controllers_interface/src/controller_manager.cpp [hunk 5]
```diff
@@ -203,10 +203,33 @@ void ControllerManager::reset()
...
+  joints_.push_back(jointHandlePtr);
+  return true;
+}
+
+bool ControllerManager::addGyroHandle(GyroHandlePtr gyroHandlePtr)
+{
+  // If we don't already have a handle with
+  // a duplicate name, we add it
+
+  for(auto gyroHandlePointer : gyros_)
```
Comment (dbking77) [inline robot_controllers_interface/src/controller_manager.cpp:226]: use underscore for variables.   For local variables don't put trailing underscore.
```
 gyro_handle_pointers
```

**Review by dbking77 — CHANGES_REQUESTED** [review]
Move setters and member variables to implementation.  Only have getters in interface.

Use underscores instead of camelCase for variables to match ROS (and google and Fetch) C++ styles.

File: robot_controllers_interface/src/controller_manager.cpp (hunk 5 above)
Comment (mikeferguson) [reply robot_controllers_interface/src/controller_manager.cpp:226]: Also, just a note - spacing doesn't follow the ROS C++ style guide (should be a space after if/for in all of these statements).

**Comment by cjds** [comment]
Going to close this out in favour of https://github.com/fetchrobotics/robot_controllers/pull/43/files

**Pull request closed without merging**